File: abc_training/E/E_Ranges_on_Tree.rs
//            E - Ranges on Tree           
// ----------------------------------------
// 問題
// https://atcoder.jp/contests/abc240/tasks/abc240_e
// ----------------------------------------

// attributes
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(non_snake_case)]
#![allow(dead_code)]
#![allow(unused_macros)]

// imports
use itertools::Itertools;
use proconio::{input, fastout, marker::{Chars, Bytes, Usize1}};

macro_rules! debug {
    ( $($val:expr),* $(,)* ) => {{
        #[cfg(debug_assertions)]
        eprintln!( concat!($(stringify!($val), " = {:?}, "),*), $($val),* );
    }};
}

// constant
const MOD1: usize = 1_000_000_007;
const MOD9: usize = 998_244_353;
const INF: usize = 1001001001001001001;

// main
fn main() {
    input! {
        N: usize,
        edges: [(Usize1, Usize1); N - 1],
    }

    let mut G = vec![vec![]; N];
    for &(u, v) in &edges {
        G[u].push(v);
        G[v].push(u);
    }

    // 区間の列
    let mut ranges = vec![(INF, INF); N];
    ranges[0].0 = 1;

    // 木DP
    rec(INF, 0, 1, &G, &mut ranges);

    // 表示
    for &(l, r) in &ranges {
        println!("{} {}", l, r);
    }
}

fn rec(p: usize, u: usize, l: usize, G: &Vec<Vec<usize>>, ranges: &mut Vec<(usize, usize)>) {
    debug!(u, &ranges);
    // 葉である場合
    if G[u].len() == 1 && G[u][0] == p {
        ranges[u].0 = l;
        ranges[u].1 = l;
    }

    let mut i = 0;
    for &v in &G[u] {
        if v != p {
            ranges[v].0 = l + i;
            rec(u, v, l + i, G, ranges);
            ranges[u].1 = ranges[v].1;
            i += 1;
        }
    }
}

File: atcoder_training/abc240/E_Ranges_on_Tree.py
#            E - Ranges on Tree           
# ----------------------------------------
# 問題
# https://atcoder.jp/contests/abc240/tasks/abc240_e
# ----------------------------------------

import sys
sys.setrecursionlimit(1000000)

N = int(input())
edges = [tuple(int(v) - 1 for v in input().split()) for _ in range(N - 1)]

# グラフの構築
G = [[] for _ in range(N)]
for u, v in edges:
    G[u].append(v)
    G[v].append(u)

# 変数の定義
x = 1
left = [-1] * N
right = [-1] * N

def rec(p, u):
    global x, left, right
    if len(G[u]) == 1 and G[u][0] == p:
        left[u] = right[u] = x
        x += 1
        return
    
    l = x
    r = x
    for v in G[u]:
        if v == p:
            continue
        rec(u, v)
        l = min(l, left[v])
        r = max(r, right[v])
    left[u] = l
    right[u] = r

rec(-1, 0)

# 表示
for i in range(N):
    print(left[i], right[i])
